src/XAD/Hessian.hpp

```diff
@@ -22,79 +22,53 @@
-{
-};
-}  // namespace detail
-
 // fwd_adj 2d vector
 template <typename T>
-std::vector<std::vector<xad::AReal<xad::FReal<T>>>> computeHessian(
-    std::vector<xad::AReal<xad::FReal<T>>> &v,
-    std::function<xad::AReal<xad::FReal<T>>(std::vector<xad::AReal<xad::FReal<T>>> &)> func,
-    xad::Tape<xad::FReal<T>> *tape)
+std::vector<std::vector<T>> computeHessian(
+    const std::vector<xad::AReal<xad::FReal<T>>> &vec,
+    std::function<xad::AReal<xad::FReal<T>>(std::vector<xad::AReal<xad::FReal<T>>> &)> foo,
+    xad::Tape<xad::FReal<T>> *tape = xad::Tape<xad::FReal<T>>::getActive())
 {
-    unsigned int domain(static_cast<unsigned int>(v.size()));
-    std::vector<std::vector<xad::AReal<xad::FReal<T>>>> matrix(
-        domain, std::vector<xad::AReal<xad::FReal<T>>>(domain, 0.0));
-    std::function<xad::AReal<xad::FReal<T>>(std::vector<xad::AReal<xad::FReal<T>>> &)> foo = func;
-
-    tape->registerInputs(v);
-
-    for (unsigned int i = 0; i < domain; i++)
-    {
-        derivative(value(v[i])) = 1.0;
-        tape->newRecording();
-
-        xad::AReal<xad::FReal<T>> y = foo(v);
-        tape->registerOutput(y);
-        value(derivative(y)) = 1.0;
-
-        tape->computeAdjoints();
-
-        for (unsigned int j = 0; j < domain; j++)
-        {
-            // std::cout << "d2y/dx" << i << "dx" << j << " = " << derivative(derivative(v[j]))
-            //           << "\n";
-            matrix[i][j] = derivative(derivative(v[j]));
-        }
-
-        derivative(value(v[i])) = 0.0;
-    }
-
+    std::vector<std::vector<T>> matrix(vec.size(), std::vector<T>(vec.size(), 0.0));
+    computeHessian(vec, foo, begin(matrix), end(matrix), tape);
     return matrix;
 }
 
 // fwd_adj iterator
 template <class RowIterator, typename T>
 void computeHessian(
-    std::vector<xad::AReal<xad::FReal<T>>> &v,
-    std::function<xad::AReal<xad::FReal<T>>(std::vector<xad::AReal<xad::FReal<T>>> &)> func,
-    xad::Tape<xad::FReal<T>> *tape, RowIterator first, RowIterator last)
+    const std::vector<xad::AReal<xad::FReal<T>>> &vec,
+    std::function<xad::AReal<xad::FReal<T>>(std::vector<xad::AReal<xad::FReal<T>>> &)> foo,
+    RowIterator first, RowIterator last,
+    xad::Tape<xad::FReal<T>> *tape = xad::Tape<xad::FReal<T>>::getActive())
 {
+    auto v(vec);
     unsigned int domain(static_cast<unsigned int>(v.size()));
-    std::function<xad::AReal<xad::FReal<T>>(std::vector<xad::AReal<xad::FReal<T>>> &)> foo = func;
 
     if (std::distance(first, last) != domain)
         throw OutOfRange("Iterator not allocated enough space");
-    static_assert(detail::has_begin<typename std::iterator_traits<RowIterator>::value_type>::value,
-                  "RowIterator must dereference to a type that implements a begin() method");
+    static_assert(
+        !xad::detail::has_begin<typename std::iterator_traits<RowIterator>::value_type>::value,
+        "RowIterator must dereference to a type that implements a begin() method");
```

Comment:
Best do the error checks before anything else (first in function). For example making a copy of the vector is not necessary if iterator is out of range.

Reply:
Well spotted, I'll do that now.